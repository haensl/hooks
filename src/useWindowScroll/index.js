import { useCallback, useState } from 'react';
import { platform } from '@haensl/services';
import { useDebounce, useOnScroll } from '../';

const useWindowScroll = (debounceMs = 25) => {
  const [scrollPosition, setScrollPosition]
    = useState(platform.scrollPosition());

  const onScroll = useCallback(() => {
    setScrollPosition(platform.scrollPosition());
  }, []);

  const onScrollDebounced = useDebounce(onScroll, debounceMs);

  useOnScroll(onScrollDebounced);

  return scrollPosition;
};

export default useWindowScroll;

