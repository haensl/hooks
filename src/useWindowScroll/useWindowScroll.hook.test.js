import React from 'react';
import { act } from 'react-dom/test-utils';
import { shallow, mount } from 'enzyme';
import useWindowScroll from './';

describe('useWindowScroll', () => {
  let TestComponent;
  let testComponent;
  let scroll;

  beforeAll(() => {
    jest.useFakeTimers();
    TestComponent = () => {
      scroll = useWindowScroll();

      if (scroll) {
        return (<span>{`${scroll.x},${scroll.y}`}</span>);
      }

      return (<span>no window scroll</span>);
    };
  });

  it('renders without crashing', () => {
    expect(shallow.bind(shallow, <TestComponent />))
      .not.toThrow();
  });

  it('renders as expected', () => {
    expect(shallow.bind(shallow, <TestComponent />))
      .toMatchSnapshot();
  });

  describe('when the window is scrolled', () => {
    beforeAll(() => {
      testComponent = mount(<TestComponent />);
    });

    beforeEach(() => {
      act(() => {
        window.scrollX = 10;
        window.scrollY = 50;
        window.dispatchEvent(new Event('scroll'));
        jest.runOnlyPendingTimers();
        testComponent.update();
      });
    });

    it('updates the window scroll position', () => {
      expect(scroll).toEqual(
        expect.objectContaining({
          x: 10,
          y: 50
        })
      );
    });

    it('renders as expected', () => {
      expect(testComponent).toMatchSnapshot();
    });
  });
});
