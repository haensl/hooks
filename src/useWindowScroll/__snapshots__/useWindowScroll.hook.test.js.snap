// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useWindowScroll renders as expected 1`] = `[Function]`;

exports[`useWindowScroll when the window is scrolled renders as expected 1`] = `
<TestComponent>
  <span>
    10,50
  </span>
</TestComponent>
`;
